src/assets/image_asset_manager.rs

```diff
@@ -0,0 +1,108 @@
+        let path: String = path.into();
+        self.image_path_queue.push(format!("{}{}", self.asset_path, path)).unwrap();
+    }
+
+    pub fn update(&mut self, device: &wgpu::Device) -> wgpu::CommandBuffer {
+        let mut encoder = device.create_command_encoder(&wgpu::CommandEncoderDescriptor {
+            label: Some("ImageUpload")
+        });
+
+        for _ in 0..self.image_path_queue.len() {
+            let path = self.image_path_queue.pop().unwrap();
+            let id = self.image_info_manager.insert(PathBuf::from(path.clone()));
+            self.image_info_manager.load(id).unwrap();
+            self.image_info_queue.push((path.clone(), id)).unwrap();
+        }
+
+        let mut image_info_still_loading = Vec::new();
+        for _ in 0..self.image_info_queue.len() {
+            let item = self.image_info_queue.pop().unwrap();
+            let image_info = self.image_info_manager.get(item.1);
+            if image_info.is_none() {
+                image_info_still_loading.push(item);
+            } else {
+                // Get path to image.
+                let full_path = PathBuf::from(&item.0);
+                let original_file_name = full_path.file_name().unwrap().to_str().unwrap();
+                let image_info: Arc<ImageInfo> = image_info.unwrap();
+                let image_path = format!("{}{}", str::replace(&item.0, original_file_name, ""), image_info.file);
+                let image_builder_id = self.image_builder_manager.insert(PathBuf::from(image_path.clone()));
+                self.image_builder_manager.load(image_builder_id).unwrap();
+                self.image_builders_queue.push((image_path, image_info.clone(), image_builder_id)).unwrap();
+            }
+        }
+
+        let mut image_builders_still_loading = Vec::new();
+        for _ in 0..self.image_builders_queue.len() {
+            let data = self.image_builders_queue.pop().unwrap();
+
+            let image_builder = self.image_builder_manager.get(data.2);
+
+            if image_builder.is_none() {
+                image_builders_still_loading.push(data);
+            } else {
+                let image_builder: Arc<ImageBuilder> = image_builder.unwrap();
+                let image = image_builder.build(&device, &mut encoder, &data.1);
+                // TODO: Store image somewhere that can be accessed by users easily.
+                log::info!("Loaded new image: {}", data.1.file);
+                self.image_storage.insert((*data.1).clone(), Arc::new(image));
+            }
+        }
+
+        for _ in 0..image_info_still_loading.len() {
+            let item = image_info_still_loading.pop().unwrap();
+            self.image_info_queue.push(item).unwrap();
+        }
+
+        for _ in 0..image_builders_still_loading.len() {
+            let item = image_builders_still_loading.pop().unwrap();
+            self.image_builders_queue.push(item).unwrap();
+        }
+
+        self.image_info_manager.maintain();
```

Comment:
Should maintain first i think.

Reply:
I was thinking that, but I wasn't sure..